book/running-wallaroo/autoscale.md

```diff
@@ -0,0 +1,83 @@
+
+{% codetabs name="Python", type="py" -%}
+```
+machida --application-module alphabet --in 127.0.0.1:7010 \
+--out 127.0.0.1:7002 --join 127.0.0.1:12500 --name w3 --worker_count 3 \
+--ponythreads 1
+
+machida --application-module alphabet --in 127.0.0.1:7010 \
+--out 127.0.0.1:7002 --join 127.0.0.1:12500 --name w4 --worker_count 3 \
+--ponythreads 1
+
+machida --application-module alphabet --in 127.0.0.1:7010 \
+--out 127.0.0.1:7002 --join 127.0.0.1:12500 --name w5 --worker_count 3 \
+--ponythreads 1
+```
+{%- language name="Go", type="go" -%}
+```
+alphabet --in 127.0.0.1:7010 --out 127.0.0.1:7002 \
+--join 127.0.0.1:12500 --name w3 --worker_count 3
+
+alphabet --in 127.0.0.1:7010 --out 127.0.0.1:7002 \
+--join 127.0.0.1:12500 --name w4 --worker_count 3
+
+alphabet --in 127.0.0.1:7010 --out 127.0.0.1:7002 \
+--join 127.0.0.1:12500 --name w5 --worker_count 3
+```
+{%- endcodetabs %}
+
+## Shrink to Fit
+
+It is also possible to remove one or more workers from a running cluster. To qualify for removal from the cluster, a worker must currently meet two criteria:
+
+1. It must have no sources placed on it.
+2. It must have no non-parallelized stateless computations placed on it.
+
+There are two ways to specify that workers should be removed from the cluster.  You can either specify the worker names or the total count of workers to be removed.  You must know the external channel address of one worker in the running cluster.  You can use the `external_sender` tool to send a shrink message to this channel. For example, if the external address is `127.0.0.1:5050` and we want to remove 2 workers from the cluster, we can run the following:
+
+```
+external_sender --type shrink --external 127.0.0.1:5050 --message 2
+```
+
+The `--message` argument is used either to pass in a count or to pass in a list of comma-separated worker names.  To request that workers `w2` and `w3` be removed, we can run:
+
+```
+external_sender --type shrink --external 127.0.0.1:5050 --message w2,w3
+```
+
+If you send in a list of worker names that includes at least one invalid worker (either because the worker does not meet the 2 criteria or because it is an invalid name) then the shrink will not happen.  If you send in a shrink count that’s higher than the number of workers meeting the 2 criteria, then Wallaroo will shrink by the maximum number of workers that do meet those criteria (which could be 0).
+
+The leaving workers will migrate state to the remaining workers and then shut down.  The remaining workers will then handle all work for the application.
+
+You can query a running cluster to get a list of workers eligible for shutdown. In order to do this, pass `?` as the argument to `--message` as below:
+
+```
+external_sender -t shrink -e 127.0.0.1:5050 -m ? -s
```

Comment:
We should use long arguments (`--type`, `--external`, etc) for these commands.

Also, zsh treats the `?` character as a special character. If you don't escape it then bad things happen. I'm not sure if other shells do this, but it makes me think that `?` may not be the best option to use here.

Reply:
This is a good point.  I've updated this to read `\?` in both places, but maybe we should change this.